### ISSUE TITLE
DOSBox rendering glitches with aspect=true and output=surface

### ISSUE THREAD

**Issue by sulix** [libsdl-org/sdl12-compat#58]
DOSBox 0.74-3 has some very strange rendering issues when aspect ratio correction is enabled, the "surface" output system (SDL 1.2 Surfaces) is selected, and a video mode with non-square pixels is being emulated.

What appears to be happening is that full-screen updates (e.g., scrolling, palette updates, etc) are not being scaled to the correct aspect ratio, leaving a blank bar at the bottom of the screen. However, partial updates (moving sprites, menu item highlights, etc), are rendered with the correct aspect ratio, and so appear further down the screen, potentially leaving trails behind them.

![Commander Keen 1 under DOSBox](https://davidgow.net/stuff/dosbox-sdl12compat.png)

I tested this on X11 and Wayland, it was easily reproducible on both. This issue is not reproducible with either the "opengl" or "overlay" (SDL 1.2 YUV overlay) output video systems, both of which seem to work perfectly under sdl12-compat. (The "ddraw" path relies pretty heavily on the hwdata field of SDL_Surface, so while I didn't test under windows, there's no realistic chance of it working.)

From a quick look at DOSBox's code, it seems to have a number of different ways it can render in "surface" mode, depending on whether the window surface is double buffered. On sdl12-compat on my systems, it uses its "blit" path, where it creates a surface and renders into that (interestingly, without locking the surface, but unlocking it afterwards), then SDL_BlitSurface()-ing it into the window surface, and calling SDL_Flip(). The non-blit path, on the other hand, locks the window surface directly and then uses either SDL_Flip() for full-screen updates and SDL_UpdateRects(), and this is the path which normal SDL 1.2 uses. It's possible that DOSBox on the "blit" codepath is just broken, of course…

**Comment by sulix** [comment]
Commit b9d3a68f37 (``Don't mark the screen surface as SDL_HWSURFACE``) breaks DOSBox rendering with ``output=surface``completely.

I get a completely black screen (or, occasionally on one of my machines, some random garbage on the display).

DOSBox [does check](https://sourceforge.net/p/dosbox/code-0/HEAD/tree/dosbox/tags/RELEASE_0_74_3/src/gui/sdlmain.cpp#l584) if the screen has ``SDL_HWSURFACE`` set, and presumably does something differently as a result.

**Comment by sulix** [comment]
Okay — I think we're actually hitting a bug in DOSBox, at least for the current case.

DOSBox:
- Works perfectly if the screen surface is neither ``SDL_HWSURFACE`` nor ``SDL_DOUBLEBUF``
- Gives the weird aspect-ratio glitches if ``SDL_HWSURFACE`` is set
- Renders nothing at all if ``SDL_DOUBLEBUF`` is set, but not ``SDL_HWSURFACE``

Basically, if ``SDL_DOUBLEBUF`` is set, DOSBox allocates a secondary "Blit" surface, which it renders to. In theory, it then will blit this to the screen, but the check for this is inside an ``SDL_MUSTLOCK(screen surface)`` check. If the  screen surface doesn't need to be locked (because it isn't an ``SDL_HWSURFACE``), then DOSBox renders to the blit surface, but never then blits it to the screen: it assumes that it had written into the screen surface directly, and uses its ``SDL_UpdateRects()`` based path.

This is clearly a bug in DOSBox — the offending code can be found [here](https://sourceforge.net/p/dosbox/code-0/HEAD/tree/dosbox/tags/RELEASE_0_74_3/src/gui/sdlmain.cpp#l861) — but it's probably worth not triggering it if we can, as DOSBox hasn't fixed it, and hasn't seen a new release in ages. (And there are several old games bundled with old DOSBox versions out there.)

I'm not sure if the original issue in this bug report is another DOSBox bug — if so, the "correct" solution here is to not have the screen surface be marked as ``SDL_DOUBLEBUF``, which matches what SDL 1.2 exposes for me, and seems to work fine.

**Comment by icculus** [comment]
Removing SDL_DOUBLEBUF is the right solution. Basically nothing did anything but software surfaces right, because SDL almost universally couldn't give you a hardware surface in the common cases.

The screen surface was meant to accurately reflect modern realities: you have to lock it (HWSURFACE) and you have to flip it a page at a time (DOUBLEBUF) but in practice it's working better to expose this as a software surface and paper over those details.

tl;dr: let's just remove SDL_DOUBLEBUF.

**Comment by sulix** [comment]
Sent PR #74 to get rid of SDL_DOUBLEBUF.

**Comment by icculus** [comment]
So is DosBox totally working now? Can we close this bug?

**Comment by sulix** [comment]
Yeah: DOSBox seems to work pretty well now. I haven't noticed any graphical issues.

I did get a crash that seemed to be audio related at one point, but haven't been able to reproduce it, and it's clearly not related to this, so I'll file a separate issue for it if it reoccurs.

**Comment by DominusExult** [comment]
@sulix when you say that DOSBox works for you, don't you get the error that all keyboard input returns "a" in DOSBox?
I'm still getting that and wonder if it is either a problem on macOS only or due to me using an outdated SDL2 lib (git of some months ago :)).

**Comment by sulix** [comment]
Hmm… keyboard input was working fine for me, but I did have ``usescancodes=false`` in my ``dosbox.conf`` file, which isn't the default.

That being said, even without it, I'm not getting "a" for everything.

Of course, DOSBox's ``usescancodes`` feature has been buggy even on real SDL 1.2, (I've had no keyboard input working at all in the past with it), so it's probably worth having it off for actual use anyway, IMHO.

In any case, let's track DOSBox keyboard issues in #84 — sdl12-compat doesn't properly translate scancodes yet anyway, which is probably (at least part of) the cause.

**Comment by icculus** [comment]
Fwiw, dosbox also worked for me, so this is likely macOS-specific.
